### ISSUE TITLE
logs are duplicated into messages when imfile is active

### ISSUE THREAD

**Issue by asarubbo** [rsyslog/rsyslog#5222]
### Environment
- rsyslog version: 8.2102.0-113.el9_2 (aka 2021.02)
- platform: CentOS 7 or RockyLinux 9

Hello, we are using rsyslog to send logs to a logstash server via `imfile` with a configuration like this:

```
$DefaultNetstreamDriver gtls
$DefaultNetstreamDriverCAFile /etc/ssl/certs/ca-bundle.crt
$ActionSendStreamDriverMode 1
$ActionSendStreamDriverAuthMode anon
$ActionQueueType Disk
$ActionQueueFileName queue_file
template(name="rsyslog"
  type="list") {
    constant(value="{")
      constant(value="\"@timestamp\":\"")     property(name="timereported" dateFormat="rfc3339")
      constant(value="\",\"@version\":\"1")
      constant(value="\",\"message\":\"")     property(name="msg" format="json")
      constant(value="\",\"sysloghost\":\"")  property(name="hostname")
      constant(value="\",\"severity\":\"")    property(name="syslogseverity-text")
      constant(value="\",\"facility\":\"")    property(name="syslogfacility-text")
      constant(value="\",\"programname\":\"") property(name="programname")
      constant(value="\",\"procid\":\"")      property(name="procid")
    constant(value="\"}\n")
}
module(load="imfile" PollingInterval="600")
input(type="imfile" File="/var/log/nginx/*log" Tag="MYTAG" Facility="local3")
local3.* @@logstash.server:PORT;rsyslog
```

By using this configuration we get the entire content of the logs configured as `input` (/var/log/nginx/*log in this case) into `/var/log/messages`.

NOTE:
The issue happens also when we comment the logstash server, so is not related to logstash or PollingInterval but just to imfile.

Is this an expected behavior? Is there a way to avoid duplication in messages?
Thanks

**Comment by davidelang** [comment]
you don't give us your full config, so it's hard to know for sure, but my guess 
is that the config file that you include that has the action to write to 
messages has a filter on it that includes local3

syslog daemons have always had the ability to do more than one thing wiht a 
given log message, so the combination of files is telling it that you want the 
logs written to messages as well as being sent remotely (this is actually a very 
common thing for people to do)

you can change the filter for messages to not include local3 (local3.none I 
believe), or if you are sure that you don't want any other processing of the 
messages from imfile, you can define a ruleset with just the actions that you 
want to have in place for that input, and add a ruleset parameter to the input() 
statement

David Lang

**Comment by asarubbo** [comment]
> you don't give us your full config, so it's hard to know for sure

Hi,

what I've already provided is `/etc/rsyslog.d/logstash.conf` apart that we have the default distro configuration, this is the CentOS 7 case:

```
$ModLoad imuxsock # provides support for local system logging (e.g. via logger command)
$ModLoad imjournal # provides access to the systemd journal
$WorkDirectory /var/lib/rsyslog
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$IncludeConfig /etc/rsyslog.d/*.conf
$OmitLocalLogging on
$IMJournalStateFile imjournal.state
*.info;mail.none;authpriv.none;cron.none                /var/log/messages
authpriv.*                                              /var/log/secure
mail.*                                                  -/var/log/maillog
cron.*                                                  /var/log/cron
*.emerg                                                 :omusrmsg:*
uucp,news.crit                                          /var/log/spooler
local7.*                                                /var/log/boot.log
```

Unless implicit or I am missing something there is no trace of local3 in the conf.

**Comment by davidelang** [comment]
*.info matches local3.info (or more severe, basically all local3 except debug)

David Lang

**Comment by asarubbo** [comment]
I understand now.

By trying other locals (0to2 and 4to7) I can avoid logging in messages if I use localN.none in the /var/log/messages configuration line, however by doing this logs are not uploaded to logstash but `rsyslog -dn` do not reports error, but I'm sure I'm still missing something.

Which Facility I can should choose to do togheter upload on logstash and avoid logging in messages?

**Comment by allamiro** [comment]
@asarubbo  To avoid logging specific facilities in /var/log/messages  you can negate a facility in the rsyslog configuration. 
 For example  if you don't want local0 messages in /var/log/messages, you can add the following:

```
*.info;mail.none;authpriv.none;cron.none;local3.none             /var/log/messages

```
Then if you want to forward to your logstash server 

```
local3.*   @@logstash.server:PORT;rsyslog

```
I hope this explains the situation

**Comment by asarubbo** [comment]
@allamiro that's what I did and the logs are not duplicated in `/var/log/messages` but by using this configuration, upload to logstash does not work (but don't ask me why..)

**Comment by rgerhards** [comment]
We are still missing the full config and the sequence of any snippets. Anyhow I wonder why you use imfile at all. Does ngnix write that log file or do these logs come in via syslog, are recorded by rsyslog and then re-read?

**Comment by asarubbo** [comment]
@rgerhards full config is at https://github.com/rsyslog/rsyslog/issues/5222#issuecomment-1713597235 and at https://github.com/rsyslog/rsyslog/issues/5222#issue-1890102134

Nginx atm is just an example (but it writes log itself), we have custom services that produce logs, and we load all to logstash.

**Comment by rgerhards** [comment]
So this is the current full config:

```
$ModLoad imuxsock # provides support for local system logging (e.g. via logger command)
$ModLoad imjournal # provides access to the systemd journal
$WorkDirectory /var/lib/rsyslog
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
### REMOVED, snippets merged: below $IncludeConfig /etc/rsyslog.d/*.conf
$DefaultNetstreamDriver gtls
$DefaultNetstreamDriverCAFile /etc/ssl/certs/ca-bundle.crt
$ActionSendStreamDriverMode 1
$ActionSendStreamDriverAuthMode anon
$ActionQueueType Disk
$ActionQueueFileName queue_file
template(name="rsyslog"
  type="list") {
    constant(value="{")
      constant(value="\"@timestamp\":\"")     property(name="timereported" dateFormat="rfc3339")
      constant(value="\",\"@version\":\"1")
      constant(value="\",\"message\":\"")     property(name="msg" format="json")
      constant(value="\",\"sysloghost\":\"")  property(name="hostname")
      constant(value="\",\"severity\":\"")    property(name="syslogseverity-text")
      constant(value="\",\"facility\":\"")    property(name="syslogfacility-text")
      constant(value="\",\"programname\":\"") property(name="programname")
      constant(value="\",\"procid\":\"")      property(name="procid")
    constant(value="\"}\n")
}
module(load="imfile" PollingInterval="600")
input(type="imfile" File="/var/log/nginx/*log" Tag="MYTAG" Facility="local3")
local3.* @@logstash.server:PORT;rsyslog
### end config snippets
$OmitLocalLogging on
$IMJournalStateFile imjournal.state
*.info;mail.none;authpriv.none;cron.none                /var/log/messages
authpriv.*                                              /var/log/secure
mail.*                                                  -/var/log/maillog
cron.*                                                  /var/log/cron
*.emerg                                                 :omusrmsg:*
uucp,news.crit                                          /var/log/spooler
local7.*                                                /var/log/boot.log
```

No other snippet inside the include dir?

**Comment by asarubbo** [comment]
> No other snippet inside the include dir?

It shouldn't have influences but there is `/etc/rsyslog.d/21-cloudinit.conf` with:

```
# Log cloudinit generated log messages to file
:syslogtag, isequal, "[CLOUDINIT]" /var/log/cloud-init.log

# comment out the following line to allow CLOUDINIT messages through.
# Doing so means you'll also get CLOUDINIT messages in /var/log/syslog
& stop
```

**Comment by rgerhards** [comment]
interesting. Comment out that `& stop` and check if things change.

Keep in mind that no matter how complex and hard-to-understand you split the config, rsyslog always only sees a single text config with all the snippets merged together. This stop is before the forwarding, so it could affect things. It's unlikely, as I would expect the condition should not match with the ngnix messages. Still worth a try.

**Comment by asarubbo** [comment]
Commented out the `& stop` but upload do not work.

I'm attaching the output of `rsyslogd -dn` after `nginx access_log` changes. I Just stripped out personal data with __UPCASE__WORDS__

[rsyslogd.txt](https://github.com/rsyslog/rsyslog/files/12606581/rsyslogd.txt)

**Comment by asarubbo** [comment]
Nevermind, please disregard my latest comment.

By commenting out `& stop` upload works for me.

It is interesting that without `local3.none` in the `/var/log/messages` line, upload works even with `& stop`.

Sorry to have missed it out.

Hint:
if it worth it, I'd suggest to create an rsyslogd option that prints the entire config (snippets) included, that rsyslogd will use. It will help for debug and may be requested in the bug reports template.

**Comment by rgerhards** [comment]
> Nevermind, please disregard my latest comment.
> 
> By commenting out `& stop` upload works for me.
> 
> It is interesting that without `local3.none` in the `/var/log/messages` line, upload works even with `& stop`.
> 
> Sorry to have missed it out.
> 
> Hint: if it worth it, I'd suggest to create an rsyslogd option that prints the entire config (snippets included), that rsyslogd will use. It will help for debug and may be requested in the bug reports template.

There is actually the `-o filename` option which does that. Often, unfortunately, people do not understand how to use it, thus in "simpler cases" I  do no longer point to it.

**Comment by rgerhards** [comment]
Side-note there are better options to do what you want to by using a dedicated ruleset. I just had not time to describe it for your case.

**Comment by asarubbo** [comment]
> Side-note there are better options to do what you want to by using a dedicated ruleset. I just had not time to describe it for your case.

Thanks for let me know, but if you can point me to an example (also with newer syntax) I will be glad to update.

**Comment by rgerhards** [comment]
It's along the lines of multiple rulesets:

https://www.rsyslog.com/doc/master/concepts/multi_ruleset.html#:~:text=1%2C%20rsyslog%20supports%20multiple%20rulesets,be%20available%20with%20all%20inputs.

In pseudocode you can do
```
input(type="imfile" File="/var/log/nginx/*log" Tag="MYTAG" Facility="local3" ruleset="nginx_forward")
ruleset(name="nginx:_forward") {
     # no filter needed, this only get's messages from the inputs where it is listed
     @@logstash.server:PORT;rsyslog
}
```

new-style for the forwarding is (pseudo):
```
action(type="omfwd" protocol="tcp" server="xxx." port="port" template="rsyslog")
``` 

check omfwd doc if I got the param names right (eg. may by "target" instead of "server")

**Comment by asarubbo** [comment]
> There is actually the `-o filename` option which does that. Often, unfortunately, people do not understand how to use it, thus in "simpler cases" I do no longer point to it.

I realized I was playing with the daemon on CentOS 7, that ships a too-old version that does not have this option -.-'



> new-style for the forwarding is (pseudo):

Thanks for the hint. I have ported my conf to the new configuration style. However I have another issue that I will post as a new ticket.

**Comment by davidelang** [comment]
On Thu, 14 Sep 2023, Rainer Gerhards wrote:

> There is actually the `-o filename` option which does that. Often, unfortunately, people do not understand how to use it, thus in "simpler cases" I  do no longer point to it.

I think -o would probably be easier to use if you could dump the config while 
rsyslog is already running. I push for -o output quite a bit.

David Lang

**Comment by rgerhards** [comment]
> I think -o would probably be easier to use if you could dump the config while rsyslog is already running. I push for -o output quite a bit. David Lang

Just to provide the reasoning again:
For a "config as it is" dump, we would need to keep the text file in core, just in case someone requests it. We could do that, but I do not really see the point that much, But, granted, a few extra KiB should not hurt much.

In any case, we would need a notification method to obtain the config than. We are already using lot's of signal, so there is not much easily left.

I guess, however, you are more after the a dump of the in-memory config representation. That's a different beast. Some issues here:

* we need to update the module interface to support that - some custom modules would not do it anyways
* we can not differentiate between what was configured and what was default - all parameters would need to be dumped as we do no know that difference
* rule sets are optimized after load - the may look highly different to what is written inside the config

In short, the output would have little in common with the user-provided config. While this may provide some insight, I guess it would heavily puzzle the user and not be really useful in config issue debugging like here (we may not even know where exactly change the setting in question in the user config). Signal problem also exists.

Just my current state of thinking.

**Comment by davidelang** [comment]
On Fri, 15 Sep 2023, Rainer Gerhards wrote:

>> I think -o would probably be easier to use if you could dump the config while rsyslog is already running. I push for -o output quite a bit. David Lang
>
> Just to provide the reasoning again:
> For a "config as it is" dump, we would need to keep the text file in core, just in case someone requests it. We could do that, but I do not really see the point that much, But, granted, a few extra KiB should not hurt much.

No, I was thinking more along the lines of how -N reads the config and reports 
possible problems and then exits, -o would not care if a copy of rsyslog is 
already running, it would just read the configs, combine them, output them and 
exit without starting anything.

> In any case, we would need a notification method to obtain the config than. We are already using lot's of signal, so there is not much easily left.
>
> I guess, however, you are more after the a dump of the in-memory config representation. That's a different beast. Some issues here:
>
> * we need to update the module interface to support that - some custom modules would not do it anyways

This is a fatal flaw, but I'm not looking for the config to be dumped some 
arbitrary time after startup. Would it still require a module interface update 
to dump the config at startup?

> * we can not differentiate between what was configured and what was default - all parameters would need to be dumped as we do no know that difference
> * rule sets are optimized after load - the may look highly different to what is written inside the config
>
> In short, the output would have little in common with the user-provided 
> config. While this may provide some insight, I guess it would heavily puzzle 
> the user and not be really useful in config issue debugging like here (we may 
> not even know where exactly change the setting in question in the user 
> config). Signal problem also exists.

I think having an option to dump the config (instead of starting, as above) 
combined with defaults and optimized (and I assume all in the new action() 
format) would be a useful option to have.

it would provide an easy 'how do I convert this config file to the new format' 
path and a 'what exactly is rsyslog doing' answer. I agree it's a more advanced 
tool that isn't suitable for everything and the -o option to just show the 
config as it's being read is a valuable earlier step

David Lang

> Just my current state of thinking.
>
>
>

**Comment by rgerhards** [comment]
> > Just to provide the reasoning again:
> > For a "config as it is" dump, we would need to keep the text file in core, just in case someone requests it. We could do that, but I do not really see the point that much, But, granted, a few extra KiB should not hurt much.
>
> No, I was thinking more along the lines of how -N reads the config and reports
> possible problems and then exits, -o would not care if a copy of rsyslog is
> already running, it would just read the configs, combine them, output them and
> exit without starting anything.

I guess I always misunderstood this. I think just merging the text
files and exiting should not be too much of an issue, but we actually
need to load the modules to do syntax checks. That might be a problem.
I'll have a look.

Rainer

**Comment by rgerhards** [comment]
@davidelang well, actually, you can do

```
rsyslogd -N1 -o config.dump
```

To make rsyslog start and shutdown immediately while creating the config dump. I guess this is sufficient?

**Comment by davidelang** [comment]
On Mon, 18 Sep 2023, Rainer Gerhards wrote:

> @davidelang well, actually, you can do
>
> ```
> rsyslogd -N1 -o config.dump
> ```
>
> To make rsyslog start and shutdown immediately while creating the config dump. I guess this is sufficient?

even if rsyslog is already running?

that sounds like a great option. Thanks for researching this.

David Lang

**Comment by rgerhards** [comment]
yup, works at least on my system. To be sure, I'll install a fresh Ubuntu VM and check it there as well. I'll report later.

**Comment by rgerhards** [comment]
Just as a side-note: the `o` option does not do much. It just outputs the text logs as they are being processed. So it's essentially a build-in text merge facility. As such, it is (or should be) compatible with anything else that is going on in rsyslog.

**Comment by rgerhards** [comment]
I can confirm that at least on Ubuntu 22.04, running `rsyslog -N1 -o config.dump` works in parallel to the running system rsyslogd (but we need root perms depending on file and module permissions). I guess this should be the same on almost all other distros. At least, everwhere, where `-N1` works.

**Comment by rgerhards** [comment]
@davidelang I have done a small write-up on the [consolidated rsyslog.conf](https://rainer.gerhards.net/2023/09/rsyslog-how-to-debug-rsyslog-conf-problems.html) and also mentioned why using many snippets is a problem in itself. I hope we can use this as future reference when requesting the consolidated config.

**Comment by davidelang** [comment]
Thanks, a thought:

"Early things happen before later things" could use some clarification.

Something along the lines of:

At Startup, Rsyslog goes through the config and implements all module load, and 
inputs, and sets up any rulesets it finds for later use (with the default 
ruleset used for all log procesing logic that's not defined as part of another 
ruleset). While processing a load message, it then goes through the appropriate 
rulesets for each log message.

Note again that as it is processing a log message, it doesn't know or care what 
snippet the log actions were in, or what input() statements may have been in the 
same file. There is a tendency to think that if you have a snippet file that 
says

input(type="file" filename="/foo/bar")
@1.2.3.4

that this means that all the logs from /foo/bar (and only those logs) will be 
sent to server 1.2.3.4, but it's really telling rsyslog to send ALL logs to 
1.2.3.4





you have an example that shows up as:

local3.* /var/log/myapp.log &
stop

the & should be on the second line

it may also be useful to show an example of a combined file.

And we may want to change the github template to explicitly ask for the output 
(and file) of rsyslogd -N1 -o /path/to/file

David Lang

**Comment by davidelang** [comment]
a possible future enhanced version of -o could create multiple files

first the combined file that we see today
second the config items evaluated at startup (i.e. everything except the third 
category)
third the config items evaluated per log message

David Lang
